### DISCUSSION TITLE
AdvisorName如何设置多个人名空格分割

### DISCUSSION THREAD

**Discussion by xzc360241369** [TheNetAdmin/zjuthesis#199] [General]
AdvisorName如何设置多个，如何加入空格，加入空格无效，使用\quad编译报错

**Comment by muzimuzhi** [comment]
dirty trick: 直接用 em dash U+2002 ` `：`AdvisorName     = 老王 老李,`。

一些解释：`kvoptions` 包虽然让 `\documentclass[<options>]{...}` 的 `<options>` 接受了 `key=value` 形式的（逗号分隔列表）参数，但文档类选项会被每个包 parse 一遍。遇到直接用 latex2e 提供的 `\ProcessOptions*` 的包时，每个 `key=value` 会被放进 `\csname ...\endcsname`，这就要求 `value` 里不能包含不可完全展开的内容，例如 `\quad`。

具体在你的例子里，使用 `\documentclass[AdvisorName=老王\quad 老李]{zjuthesis}` 时，会导致 `\csname AdvisorName=老王\hskip1em\relax 老李\endcsname`，这里的 `\hskip` 是不可展开的，从而报错
```tex
! Missing \endcsname inserted.
<to be read again> 
                   \hskip 
```

不够了解 `kvoptions` 和 latex2e 处理文档类/宏包选项的内部，给不出更好的方案。

**Reply by Chenyanlu99** [reply]
我直接复制你的U+2002空格可以成功编译，但是在ColaboratorName中有两个导师，两个导师的名字和职称中间加上空格后导致这一行的行间距特别大。这个可以修改吗？
ColaboratorName = XX 教授、XXX 研究员,
![image](https://github.com/TheNetAdmin/zjuthesis/assets/55088746/54e0897d-d0ec-4c5e-bdca-dbf74541d52c)

**Reply by muzimuzhi** [reply]
感觉是字数较多，超过允许的表格列宽了。比如我把下面第一份代码片段 122 行里的 `\begin{tabularx}{.7\textwidth}` 改成 `\begin{tabularx}{.7\textwidth}` 后，结果就正常了：
![image](https://github.com/TheNetAdmin/zjuthesis/assets/6376638/3567cb66-b16e-4aaa-a083-6d87a0925cf3)

改宽一点不涉及「不遵照学校/学院样式吧」 [doge]；建议提问时总是提供一个例子，方便他人复现。
https://github.com/TheNetAdmin/zjuthesis/blob/9476151fc2815b3d42a6613d38286ad8a379e704/page/graduate/cover-chn.tex#L122-L133
https://github.com/TheNetAdmin/zjuthesis/blob/9476151fc2815b3d42a6613d38286ad8a379e704/page/graduate/doctor/opteng/cover-chn.tex#L69-L74

**Reply by Chenyanlu99** [reply]
谢谢，按照你的方法，问题确实解决了！
遇到问题的那一句是zjuthesis.tex的第38行`ColaboratorName = XX 教授、XXX 研究员`
似乎修改[zjuthesis/page/graduate/cover-chn.tex](https://github.com/TheNetAdmin/zjuthesis/blob/9476151fc2815b3d42a6613d38286ad8a379e704/page/graduate/cover-chn.tex#L122-L133)中的\begin{tabularx}{.6\textwidth}{>{\fangsong}l >{\fangsong}X<{\centering}}为\begin{tabularx}{.7\textwidth}{>{\fangsong}l >{\fangsong}X<{\centering}}即可，不需要对[zjuthesis/page/graduate/doctor/opteng/cover-chn.tex](https://github.com/TheNetAdmin/zjuthesis/blob/9476151fc2815b3d42a6613d38286ad8a379e704/page/graduate/doctor/opteng/cover-chn.tex#L69-L74)进行操作

**Comment by muzimuzhi** [comment]
一个可复现的例子

```tex
\begin{filecontents}{myclass.cls}
\LoadClass{article}
\RequirePackage{kvoptions}
\DeclareStringOption{key}
\end{filecontents}

\begin{filecontents}{mypkg.sty}
\ProcessOptions*
\end{filecontents}

\documentclass[key=\quad]{myclass}
\usepackage{mypkg}

\begin{document}
content
\end{document}
```

得到报错
```tex
[...] (./mypkg.sty
! Missing \endcsname inserted.
<to be read again> 
                   \hskip 
l.5 \ProcessOptions*
```

**Comment by TheNetAdmin** [comment]
或者另一个土办法就是在所有使用到 AdvisorName 的地方，hard code你的内容，比如在封面页里hard code你的多个导师名字，这样应该就能用各种空白分隔符了

**Comment by Sec-ant** [comment]
https://github.com/TheNetAdmin/zjuthesis/discussions/199#discussioncomment-1471655 提出的 `U+2002` 空格宽度比较大，如果想要和普通空格保持一致，可以使用 `no-break space` `U+00A0` 这个非断行空格，LaTeX 语法是 `~`，测试过是可以正常编译的：

```latex
AdvisorName      = 导师A~导师B,
```
